pkg/tenantctl/controller.go

```diff
@@ -157,219 +167,394 @@ func (c *Controller) processNextWorkItem() bool {
+		if _, err := c.updateStatus(tenant); err != nil {
 			return err
 		}
-		tenantRoleName := config.GetTenantRoleName()
+	}
 
-		//TODO: handle updateProgress errors
-		tenant, _ = c.updateProgress(tenant, api.TenantProgressCreateNamespace)
-		namespaceName, err = c.createNamespace(tenant)
-		if err != nil {
-			return c.handleError(tenant, err, api.TenantResultErrorContent)
-		}
-		log.Printf("Create namespace successful for %s", namespaceName)
+	if reconcileErr != nil {
+		return reconcileErr
+	}
 
-		tenant, _ = c.updateProgress(tenant, api.TenantProgressGetServiceAccount)
-		account, err = c.getServiceAccount(tenant, defaultServiceAccountName)
-		if err != nil {
-			return c.handleError(tenant, err, api.TenantResultErrorInfra)
-		}
+	c.updateMetrics()
+	c.syncCount++
+	return nil
+}
 
-		tenant, _ = c.updateProgress(tenant, api.TenantProgressAddRoleBinding)
-		var roleBinding *v1beta1.RoleBinding
-		roleBinding, err = c.addRoleBinding(account, tenant, tenantRoleName)
-		if err != nil {
-			return c.handleError(tenant, err, api.TenantResultErrorInfra)
-		}
-		log.Printf("Created Role Binding '%s' in namespace '%s'", roleBinding.GetName(), namespaceName)
+func (c *Controller) reconcile(config clientConfig, tenant *api.Tenant) (err error) {
+	if tenant.Status.TenantNamespaceName == "" {
+		err = c.reconcileUninitialized(config, tenant)
+	} else {
+		err = c.reconcileInitialized(config, tenant)
+	}
+	return
+}
 
-		tenant, _ = c.updateProgress(tenant, api.TenantProgressFinalize)
-		tenant.Status.Result = api.TenantResultSuccess
-		tenant.Status.Message = "Tenant namespace successfully prepared"
-		if tenant, err = c.updateStatus(tenant); err != nil {
-			return err
+func (c *Controller) reconcileUninitialized(config clientConfig, tenant *api.Tenant) error {
+	c.logPrintln(tenant, "tenant not initialized yet")
+
+	nsName, err := c.createTenantNamespace(config, tenant)
+	if err != nil {
+		condMsg := fmt.Sprintf("Failed to create the tenant namespace.")
+		tenant.Status.SetCondition(&knativeapis.Condition{
+			Type:    knativeapis.ConditionReady,
+			Status:  corev1.ConditionFalse,
+			Reason:  api.StatusReasonFailed,
+			Message: condMsg,
+		})
+		return err
+	}
+
+	_, err = c.syncRoleBinding(tenant, nsName, config)
+	if err != nil {
+		condMsg := fmt.Sprintf("Failed to create the tenant namespace.")
+		tenant.Status.SetCondition(&knativeapis.Condition{
+			Type:    knativeapis.ConditionReady,
+			Status:  corev1.ConditionFalse,
+			Reason:  api.StatusReasonFailed,
+			Message: condMsg,
+		})
+		c.rollbackTenantNamespace(nsName, tenant, config) // clean-up ignoring error
+		return err
+	}
+
+	tenant.Status.TenantNamespaceName = nsName
+
+	tenant.Status.SetCondition(&knativeapis.Condition{
+		Type:   knativeapis.ConditionReady,
+		Status: corev1.ConditionTrue,
+	})
+
+	return nil
+}
+
+func (c *Controller) reconcileInitialized(config clientConfig, tenant *api.Tenant) error {
+	c.logPrintln(tenant, "tenant is initialized already")
```

Comment:
A re-sync of unmodified objects should happen every 5 minutes. If this is true, the log volume is acceptable and we should keep this log entry as it helps to understand the flow in case we need to analyze problems.

I'll check this.

Reply:
Re-sync is performed every minute.

Reply:
We should keep the logging as it is and gain some experience in production. If it turns out that the information has no practical value, we can remove it.